### ISSUE TITLE
False negative: missing global-buffer-overflow in ASAN

### ISSUE THREAD

**Issue by shao-hua-li** [llvm/llvm-project#54667]
For the following code, `clang -fsanitize=address -O3` reports nothing for the clear global-buffer-overflow in `a[i]` and `a[x]`. If you change the declaration from `a[5]` to `a[4]`, it would then report global-buffer-overflow..
```shell
$cat a.c
int a[5];

int foo(int x) {
    for (int i=0;i<=5;i++)
        a[i]++;
    return a[x];
}

int main(){
    __builtin_printf("foo = %d\n", foo(5));
}
$
$clang-tk -fsanitize=address -w -O3 a.c; ./a.out
foo = 1
$
```
```shell
$cat a.c
int a[4];

int foo(int x) {
    for (int i=0;i<=5;i++)
        a[i]++;
    return a[x];
}

int main(){
    __builtin_printf("foo = %d\n", foo(5));
}
$
$clang-tk -fsanitize=address -w -O3 a.c; ./a.out
==1==ERROR: AddressSanitizer: global-buffer-overflow on address
...
$
```
